jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -40,33 +45,59 @@ abstract class CodeWhispererCodeCoverageTracker(
         get() = totalTokens.get()
     val acceptedRecommendationsCount: Int
         get() = rangeMarkers.size
+    private val isActive: AtomicBoolean = AtomicBoolean(false)
     private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
     private val isShuttingDown = AtomicBoolean(false)
     private var startTime: Instant = Instant.now()
 
-    init {
+    fun activateTrackerIfNotActive() {
+        if (!isTelemetryEnabled() || isActive.get()) return
         val conn = ApplicationManager.getApplication().messageBus.connect()
         conn.subscribe(
             CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
             object : CodeWhispererUserActionListener {
                 override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, rangeMarker: RangeMarker) {
                     if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
                     rangeMarkers.add(rangeMarker)
+                    val originalRecommendation = extractRangeMarkerString(rangeMarker)
+                    originalRecommendation?.let {
+                        rangeMarker.putUserData(KEY_REMAINING_RECOMMENDATION, it)
+                    }
                 }
             }
         )
+        startTime = Instant.now()
+        isActive.set(true)
```

Comment:
should be set before listener is subscribed

or just mark the entire method as synchronized so you don’t have to worry about any of that

Reply:
hi @rli, just for my own education, could you explain a bit more what's the concern here if we set `isActive = true` after subscription of listener? thanks!